File: src/Bluebutton.m
// Bluebutton.m
#import <UIKit/UIKit.h>
#import <objc/runtime.h>
#import <objc/message.h>

void swizzled_viewDidAppear(id self, SEL _cmd, BOOL animated) {
    ((void (*)(id, SEL, BOOL))objc_msgSend)(self, @selector(original_viewDidAppear:), animated);
    
    UIViewController *vc = (UIViewController *)self;
    
    UIButton *tweakButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [tweakButton setTitle:@"Tweaked! 🎉" forState:UIControlStateNormal];
    tweakButton.backgroundColor = [UIColor systemBlueColor];
    [tweakButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    tweakButton.layer.cornerRadius = 10;
    tweakButton.translatesAutoresizingMaskIntoConstraints = NO;
    [tweakButton addTarget:vc action:@selector(tweakButtonTapped) forControlEvents:UIControlEventTouchUpInside];
    
    [vc.view addSubview:tweakButton];
    
    [NSLayoutConstraint activateConstraints:@[
        [tweakButton.centerXAnchor constraintEqualToAnchor:vc.view.centerXAnchor],
        [tweakButton.bottomAnchor constraintEqualToAnchor:vc.view.safeAreaLayoutGuide.bottomAnchor constant:-20],
        [tweakButton.widthAnchor constraintEqualToConstant:150],
        [tweakButton.heightAnchor constraintEqualToConstant:44]
    ]];
}

void tweakButtonTapped(id self, SEL _cmd) {
    UIViewController *vc = (UIViewController *)self;
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Hello!" message:@"ObjC tweak button tapped!" preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
    [vc presentViewController:alert animated:YES completion:nil];
}

__attribute__((constructor))
static void initialize() {
    NSLog(@"[Tweak] Loading...");
    
    class_addMethod([UIViewController class], @selector(tweakButtonTapped), (IMP)tweakButtonTapped, "v@:");
    
    Method originalMethod = class_getInstanceMethod([UIViewController class], @selector(viewDidAppear:));
    
    class_addMethod([UIViewController class], @selector(original_viewDidAppear:), method_getImplementation(originalMethod), method_getTypeEncoding(originalMethod));
    
    class_replaceMethod([UIViewController class], @selector(viewDidAppear:), (IMP)swizzled_viewDidAppear, method_getTypeEncoding(originalMethod));
    
    NSLog(@"[Tweak] Loaded!");
}

File: src/Bluebutton.swift
// Tweak.swift
// Simple Swift Tweak - Adds a button to every view controller

import Foundation
import UIKit
import ObjectiveC

extension UIViewController {
    @objc dynamic func swizzled_viewDidAppear(_ animated: Bool) {
        self.swizzled_viewDidAppear(animated)
        
        // Create a button
        let tweakButton = UIButton(type: .system)
        tweakButton.setTitle("Tweaked! 🎉", for: .normal)
        tweakButton.backgroundColor = .systemBlue
        tweakButton.setTitleColor(.white, for: .normal)
        tweakButton.layer.cornerRadius = 10
        tweakButton.translatesAutoresizingMaskIntoConstraints = false
        tweakButton.addTarget(self, action: #selector(tweakButtonTapped), for: .touchUpInside)
        
        self.view.addSubview(tweakButton)
        
        NSLayoutConstraint.activate([
            tweakButton.centerXAnchor.constraint(equalTo: self.view.centerXAnchor),
            tweakButton.bottomAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.bottomAnchor, constant: -20),
            tweakButton.widthAnchor.constraint(equalToConstant: 150),
            tweakButton.heightAnchor.constraint(equalToConstant: 44)
        ])
    }
    
    @objc func tweakButtonTapped() {
        let alert = UIAlertController(title: "Hello!", message: "Swift tweak button tapped!", preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .default))
        self.present(alert, animated: true)
    }
}

func swizzle() {
    let original = #selector(UIViewController.viewDidAppear(_:))
    let swizzled = #selector(UIViewController.swizzled_viewDidAppear(_:))
    
    guard let originalMethod = class_getInstanceMethod(UIViewController.self, original),
          let swizzledMethod = class_getInstanceMethod(UIViewController.self, swizzled) else {
        return
    }
    
    method_exchangeImplementations(originalMethod, swizzledMethod)
}

@_cdecl("swift_tweak_init")
public func initTweak() {
    NSLog("[Tweak] Loading...")
    swizzle()
    NSLog("[Tweak] Loaded!")
}
